tests/PHPStan/Analyser/data/bug-3789.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Bug3789;
+
+use function PHPStan\Testing\assertType;
+
+/**
+ * @param array{string, string, string} $haystack
+ */
+function doFoo(string $needle, array $haystack): void {
+	assertType('0|1|2|false', array_search($needle, $haystack, true));
+	assertType('0|1|2|false', array_search('foo', $haystack, true));
+
+	$haystack[1] = 'foo';
+	assertType('0|1|2|false', array_search($needle, $haystack, true));
+	assertType('0|1|2', array_search('foo', $haystack, true));
```

Comment:
please also add test-cases for the non-strict variant (leaving the 3rd param to `false`)

Reply:
I added test cases.